app/process/recovery/recovery.go

```diff
@@ -235,61 +234,61 @@ func (r Recovery) recoverEthereumTXMessage(tm encoding.TopicMessage) error {
-	// Iterate all keys
-	// 	If validator has not signed -> sign and submit
-	//for txn, txnSignatures := range unprocessed {
-	//hasSubmittedSignature, ctm := r.hasSubmittedSignature(txn, txnSignatures)
-	//
-	//if !hasSubmittedSignature {
-	//	r.logger.Infof("Validator has not yet submitted signature for Transaction with ID [%s]. Proceeding now...", txn)
-	//	// TODO
-	//	err = r.transfersService.VerifyFee(ctm)
-	//	if err != nil {
-	//		r.logger.Errorf("Fee validation failed for TX [%s]. Skipping further execution", transferMsg.TransactionId)
-	//	}
-	//
-	//	signature, err := r.transfersService.ValidateAndSignTxn(ctm)
-	//	if err != nil {
-	//		r.logger.Errorf("Failed to Validate and Sign TransactionID [%s]. Error [%s].", txn, err)
-	//	}
-	//
-	//	_, err = r.transfersService.HandleTopicSubmission(ctm, signature)
-	//	if err != nil {
-	//		return errors.New(fmt.Sprintf("Could not submit Signature [%s] to Topic [%s] - Error: [%s]", signature, r.topicID, err))
-	//	}
-	//	r.logger.Infof("Successfully Validated")
-	//}
-	//}
+	unprocessedMessages, err := r.messagesRepo.GetUnprocessedMessages()
+	if err != nil {
+		r.logger.Fatalf("Failed to get all unprocessedMessages messages. Error: %s", err)
+		return err
+	}
+
+	signatureMessagesMap := make(map[*TransferMessageKey][]string)
+
+	for _, txnMessage := range unprocessedMessages {
+		key := &TransferMessageKey{
+			TransactionId: txnMessage.TransactionId,
+			EthAddress:    txnMessage.EthAddress,
+			Amount:        txnMessage.Amount,
+			Fee:           txnMessage.Fee,
+			GasPriceWei:   txnMessage.GasPriceWei,
+			Erc20Address:  txnMessage.ERC20ContractAddress,
+		}
+		signatureMessagesMap[key] = append(signatureMessagesMap[key], txnMessage.Signature)
+	}
+
+	for txn, txnSignatures := range signatureMessagesMap {
+		hasSubmittedSignature, topicMessage := r.hasSubmittedSignature(txn, txnSignatures)
+
+		if !hasSubmittedSignature {
+			err = r.transfers.PrepareAndSubmitToTopic(topicMessage)
+			if err != nil {
+				r.logger.Errorf("Failed to prepare and submit Transaction ID [%s] to the HCS Topic. Error [%s].", txn.TransactionId, err)
+				return err
+			}
+		}
+	}
 	return nil
 }
 
-func (r Recovery) hasSubmittedSignature(data joined.CTMKey, signatures []string) (bool, *validatorproto.TransferMessage) {
-	//ctm := &validatorproto.TransferMessage{
-	//	TransactionId: data.TransactionId,
-	//	EthAddress:    data.EthAddress,
-	//	Amount:        data.Amount,
-	//	Fee:           data.Fee,
-	//	GasPriceGwei:  data.GasPriceGwei,
-	//}
-	//
-	//signature, err := r.transfersService.ValidateAndSignTxn(ctm)
-	//if err != nil {
-	//	r.logger.Errorf("Failed to Validate and Sign TransactionID [%s]. Error [%s].", data.TransactionId, err)
-	//}
-	//
-	//for _, s := range signatures {
-	//	if signature == s {
-	//		return true, nil
-	//	}
-	//}
-	return false, nil
+func (r Recovery) hasSubmittedSignature(data *TransferMessageKey, signatures []string) (bool, *encoding.TopicMessage) {
```

Comment:
This function seems little off to me. It implies that it does something very trivial and returns `true/false` but it actually signs the authorization message, checks whether it is contained or not and returns the formatted signature message otherwise. 

I think that it would be better if
1. we do the signing in the context of the `processUnfinishedOperations` function in the `range` where you iterate the `signatureMessagesMap`, 
2. call this function to check whether the signature is contained in the array. The function simply returns true/false
3. If `false` do the `encoding.NewSignatureMessage` (but again, in the scope of `processUnfinishedOperations`) and not here